### PR TITLE
fixes #3697 feat(nimbus): send overview mutation when form submitted

### LINKED ISSUE #3697: `PageEditOverview` "Save" experiment edits (GQL mutation) + "Next"
1) Add a "Save" and "Next" button at the bottom of this page.
2) On "Save" click, combine any changed fields with data from EXP-302 to send the GQL mutation up to the server to save the experiment.
3) Keep track of if the user makes any changes on a field to:
3a) Activate the "Save" button (default state should be disabled)
3b) disable the "Next" button (default state should be enabled)
3c) Use the available hook to alert the user if they attempt to navigate away from the page via the sidebar with unsaved changes
4) "Next" should take the user to the `/branches` page

┆Issue is synchronized with this [Jira Task](https://jira.mozilla.com/browse/EXP-463)
┆Issue Number: EXP-463

### PR DESCRIPTION
**Pull request by jodyheavener**
This PR:

- Adds the mutation to update an experiment
  - This updated generated types
- Updates the shared FormOverview to enable Save and Next based on the status of the fields
  - Save will only be enabled if there are changes
  - Next will only be enabled if there are no changes (initial state and after save)
- Updates the shared FormOverview to use config-based Application dropdown (eliminating it from #3696)
- Because why not, I can remove this if desired: enables the Cancel button in the New Experiment form (eliminating it from #3696)
- Warns you on the New and Edit pages if you make changes and attempt to leave the page
- Moves some test mutation mocking over to the shared mocks file
- Adds missing Storybook for FormOverview with experiment mock data
- Updates tests

Some of these tests were a little clunky to get to 100%, so feel free to shame me if I've done anything super unclear.

### CODE REVIEW

File: app/experimenter/nimbus-ui/src/components/FormOverview/index.test.tsx
Comment (lmorchard) [inline app/experimenter/nimbus-ui/src/components/FormOverview/index.test.tsx:234]: Not going to get the runtime app config in a test and constants won't do - this should be mocked here

Comment (jodyheavener) [reply app/experimenter/nimbus-ui/src/components/FormOverview/index.test.tsx:234]: Good point, with #3851 now merged the mocked cache now has a mock config loaded that should be useful.

**Comment by jodyheavener** [comment]
@lmorchard While this is mostly still reviewable I'm going to hold until #3849 is merged and then update to reflect the per-compoment mock structure you have set up.

File: app/experimenter/nimbus-ui/src/components/FormOverview/index.tsx
Comment (LZoog) [inline app/experimenter/nimbus-ui/src/components/FormOverview/index.tsx:34]: Hmm, I'm not sure this work as expected. I can make an edit on the overview page and click on "Branches" to navigate away after I've made an edit without an alert showing.

File: app/experimenter/nimbus-ui/src/components/PageEditOverview/index.tsx
Comment (LZoog) [inline app/experimenter/nimbus-ui/src/components/PageEditOverview/index.tsx:61]: I think we want to remove L61 or rework it somehow. When I make an edit and click "Save," the old value is repopulated in the field. If we just remove it though, the "Save" button doesn't disable with "Next" re-enabled.

File: app/experimenter/nimbus-ui/src/components/PageEditOverview/index.tsx
Comment (LZoog) [inline app/experimenter/nimbus-ui/src/components/PageEditOverview/index.tsx:33]: We need a mock in Storybook for this not to error, though it would be also be nice to have a state besides "basic" to show an error state.

image

File: app/experimenter/nimbus-ui/src/components/FormOverview/index.tsx
Comment (jodyheavener) [reply app/experimenter/nimbus-ui/src/components/FormOverview/index.tsx:34]: That's fair, this appears to be an issue with how it works with React Router (maybe it doesn't send the `beforeunload` event). It works reliably with true browser navigation for me, though. We'll probably want to break this out into a bug issue.

File: app/experimenter/nimbus-ui/src/components/PageEditOverview/index.tsx
Comment (jodyheavener) [reply app/experimenter/nimbus-ui/src/components/PageEditOverview/index.tsx:33]: Sure I can add that. FWIW I'm on the fence about this, because `FormOverview` does have its own stories (including with existing experiment data and error states), so we're really looking to demonstrate the page layout and not necessarily the form. But I can see how for completeness you should be able to successfully hit Save. Maybe we don't need all the error states though?

Comment (LZoog) [reply app/experimenter/nimbus-ui/src/components/PageEditOverview/index.tsx:33]: Yeah, I feel like things under `Pages` should work somewhat well like clicking "Save" either sends a successful mutation or simply links to `PageEditOverview` (the latter probably being easier). It'd be nice for one of the pages to show what the error state looks like, anything extra is probably bonus points 🤷

File: app/experimenter/nimbus-ui/src/components/FormOverview/index.tsx
Comment (LZoog) [reply app/experimenter/nimbus-ui/src/components/FormOverview/index.tsx:34]: WFM to do in a follow up. We can ask Ana how imperative it is that they can't navigate through the sidebar, we may want to add it to the "create/edit flow" epic and if what you've got is mostly sufficient we can file an issue in the "UX nice-to-haves".

File: app/experimenter/nimbus-ui/src/components/PageEditOverview/index.tsx
Comment (jodyheavener) [reply app/experimenter/nimbus-ui/src/components/PageEditOverview/index.tsx:61]: Fixed

File: app/experimenter/nimbus-ui/src/components/FormOverview/index.tsx
Comment (jodyheavener) [reply app/experimenter/nimbus-ui/src/components/FormOverview/index.tsx:34]: Issue filed: https://jira.mozilla.com/browse/EXP-545

File: app/experimenter/nimbus-ui/src/components/PageEditOverview/index.tsx
Comment (jodyheavener) [reply app/experimenter/nimbus-ui/src/components/PageEditOverview/index.tsx:27]: To recap our Slack discussion, it doesn't appear that there's a great approach here without pulling things apart a bit.

> clicking "Save" either sends a successful mutation

While it's possible to create a mock for this, the mutation input needs to match the form input, otherwise it won't get picked up. So a mock would immediately become invalidated once a user starts typing in any of the fields.

> like on the PageNew page, simply links to PageEditOverview

The Save button, by way of `FormOverview`, actually already links to this component! But I think Storybook intercepts this as already being on that story, so it does nothing and still results in the error being shown.

There might be other approaches that I'm not thinking of at the moment, but it sounds like we might want to follow up on this to avoid blocking? Is that correct?

File: app/experimenter/nimbus-ui/src/components/PageEditOverview/index.tsx
Comment (lmorchard) [inline app/experimenter/nimbus-ui/src/components/PageEditOverview/index.tsx:70]: Hmm, feeling a little leery of this bit of ref-based state management. But, I can't think of any clear problems this would cause.

**Review by lmorchard — APPROVED** [review]
Looks like a couple of follow-ups to pursue, but pretty much LGTM unless lzoog has further comments

File: app/experimenter/nimbus-ui/src/components/PageEditOverview/index.tsx
Comment (jodyheavener) [reply app/experimenter/nimbus-ui/src/components/PageEditOverview/index.tsx:70]: I only did this so I could define `onFormSubmit` outside of the render; but happy remove this and move it back in?

File: app/experimenter/nimbus-ui/src/components/FormOverview/index.test.tsx
Comment (LZoog) [inline app/experimenter/nimbus-ui/src/components/FormOverview/index.test.tsx:95]: Ohh, you know I think we actually want to allow saving pieces of an experiment without filling out required fields. I think a user should be able to leave "Public Description" blank if they wanted to only update say, the hypothesis, and save that, to update the public description later.

However, we could double check this with Ana and maybe do this in a follow up as I'm not _entirely_ certain of this.

**Review by LZoog — APPROVED** [review]
LGTM as well, might need to double check with Ana on intended UX for one point but nothing blocking👍

File: app/experimenter/nimbus-ui/src/components/FormOverview/index.test.tsx
Comment (jodyheavener) [reply app/experimenter/nimbus-ui/src/components/FormOverview/index.test.tsx:49]: @LZoog pointed out to me that VS Code will helpfully place a lil dot underneath `await`s that do not need it!

File: app/experimenter/nimbus-ui/src/components/FormOverview/index.test.tsx
Comment (jodyheavener) [reply app/experimenter/nimbus-ui/src/components/FormOverview/index.test.tsx:95]: Updated to make optional

File: app/experimenter/nimbus-ui/src/components/PageEditOverview/index.tsx
Comment (lmorchard) [reply app/experimenter/nimbus-ui/src/components/PageEditOverview/index.tsx:27]: Oof, speaking of the PageNew story: I just noticed that it looks like this commit broke it.

* Before: https://storage.googleapis.com/mozilla-storybooks-experimenter/commits/265a366028dce040318162dd5fa0459c45e41e69/nimbus-ui/index.html?path=/story/pages-new--basic

* After: https://storage.googleapis.com/mozilla-storybooks-experimenter/commits/5797b0f1eaeb53adb68951e40cf0f2715cc0fe8f/nimbus-ui/index.html?path=/story/pages-new--basic

**Pull request merged**
